experiments/experiment_prototype.py

```diff
@@ -1,385 +1,1111 @@
+        "txfreq", "rxfreq", "clrfrqrange" 
+        
+        :param: exp_slice: a slice to setup
+        :return: complete_slice : a checked slice with all defaults
+        """
+
+        complete_slice = copy.deepcopy(exp_slice)
+
+        # None values are useless to us - if they do not exist we know they are None.
+        for key, value in complete_slice.items():
+            if value is None:
+                complete_slice.pop(key)
+
+        try:
+            assert 'rxfreq' or 'txfreq' or 'clrfrqrange' in complete_slice.keys()
+        except AssertionError:
+            errmsg = 'An rxfreq, txfreq, or clrfrqrange must be Specified in a Slice'
+            # TODO log
+            raise ExperimentException(errmsg, exp_slice)
+
+        self.set_slice_identifiers(complete_slice)
+        self.check_slice_specific_requirements(complete_slice)
+        self.check_slice_minimum_requirements(complete_slice)
+        complete_slice = self.set_slice_defaults(complete_slice)
+        # set_slice_defaults will check for any missing values that should be given a default and
+        # fill them.
+
+        errors = self.check_slice(complete_slice)
+
+        if errors:
+            raise ExperimentException(errors)
+
+        return complete_slice
+
+    def self_check(self):
         """
         Check that the values in this experiment are valid
         """
 
-        if (self.cponum<1):
-            errmsg="Error: No objects in control program"
-            sys.exit(errmsg) # REVIEW 6 Add a todo for error handling. Perhaps use exceptions instead.
+        if self.num_slices < 1:
+            errmsg = "Error: Invalid num_slices less than 1"
+            raise ExperimentException(errmsg)
 
-        #TODO: somehow check if self.cpid is not unique
+        # TODO: somehow check if self.cpid is not unique - incorporate known cpids from git repo?
 
-        for cpo in range(self.cponum):
-            selferrs=selfcheck_piece(self.cpo_list[cpo],self.config)
-            if (not selferrs): # REVIEW #39 unneeded parenthesis
+        for a_slice in self.slice_ids:
+            selferrs = self.check_slice(self.slice_dict[a_slice])
+            if not selferrs:
                 # If returned error dictionary is empty
                 continue
-            errmsg="Self Check Errors Occurred with Object Number : {} \nSelf \
-                Check Errors are : {}".format(cpo, selferrs)
-            sys.exit(errmsg) # REVIEW 6 Add a todo for error handling. Perhaps use exceptions instead.
-        else: # no break (exit)
-            print "No Self Check Errors. Continuing..."
+            errmsg = "Self Check Errors Occurred with slice Number : {} \nSelf \
+                Check Errors are : {}".format(a_slice, selferrs)
+            raise ExperimentException(errmsg)
+
+        print("No Self Check Errors. Continuing...")
+
         return None
 
+    def check_slice(self, exp_slice):
+        """
```

Comment:
For docstrings, in general, if the function ever has the possibility of raising an exception you can use the :raise: docstring tag like so:

"""
:raise ExperimentException: When blah blah blah